File: Java/Data Structures & Algorithms Module - Java/Week10/BSTAttempt3/BST.java
/*
BST.java
Billy Cussen
28/11/2020
*/
public class BST<T extends Comparable<T>> implements BSTInterface<T>{

    Node<T> root;

    @Override
    public void insert(T element){
        if(isEmpty()){
            this.root = new Node<T>(element);
        }
        insertHelper(this.root, element);
    }

    private void insertHelper(Node<T> current, T element){
        if(element.compareTo(current.element)==-1){
            if(current.left==null){
                current.left = new Node<T>(element);
            } else{
                insertHelper(current.left,element);
            }
        } else{
            if(current.right==null){
                current.right = new Node<T>(element);
            } else{
                insertHelper(current.right,element);
            }
        }
    }

    @Override
    public void inOrder(){
        if(isEmpty()){
            throw new NullPointerException("Can't do in order traversal, BST is Empty!");
        } else{
            System.out.println("***BST IN ORDER TRAVERSAL***");
            inOrderHelper(this.root);
            System.out.println("***END***");
        }
    }

    private void inOrderHelper(Node<T> current){
        if(current==null){
            return;
        }

        inOrderHelper(current.left);
        System.out.println(current);
        inOrderHelper(current.right);
    }

    @Override
    public T remove(T element){
        if(isEmpty()){
            throw new NullPointerException("Can't Remove "+element+", the BST is Empty!");
        }
        return removeHelper(this.root, element);
    }

    private T removeHelper(Node<T> current, T element){
        return null;        
    }

    @Override
    public boolean isEmpty(){
        return this.root == null;
    }

    @Override
    public boolean contains(T element){
        if(isEmpty()){
            throw new NullPointerException("Can't check if BST contains "+element+", BST is Empty!");
        } 
        return containsHelper(this.root, element);
    }

    private boolean containsHelper(Node<T> current, T element){
        if(current == null){
            return false;
        }

        if(element.compareTo(current.element)==0){
            return true;
        } else{
            return containsHelper(current.left,element)||containsHelper(current.right,element);
        }
    }

    @Override
    public int size(){
        if(isEmpty()){
            throw new NullPointerException("Can't get Size, BST is Empty!");
        }
        return sizeHelper(this.root);
    }

    private int sizeHelper(Node<T> current){
        if(current==null){
            return 0;
        }
        return 1 + sizeHelper(current.left)+sizeHelper(current.right);
    }

    @Override
    public int depth(){
        if(isEmpty()){
            throw new NullPointerException("Can't get Depth, BST is Empty!");
        }
        return depthHelper(this.root);
    }

    private int depthHelper(Node<T> current){
        if(current==null){
            return 0;
        }

        int leftSize = depthHelper(current.left);
        int rightSize = depthHelper(current.right);

        if(leftSize>rightSize){
            return 1+leftSize;
        } else{
            return 1+rightSize;
        }
    }

    @Override
    public T max(){
        if(isEmpty()){
            throw new NullPointerException("Can't get Max, BST is Empty!");
        } 
        return maxHelper(this.root);
    }

    private T maxHelper(Node<T> current){
        if(current==null){
            return null;
        }

        if(current.right==null){
            return current.element;
        } else{
            return maxHelper(current.right);
        }
    }

    @Override
    public T min(){
        if(isEmpty()){
            throw new NullPointerException("Can't get Min, BST is Empty!");
        }
        return minHelper(this.root);
    }

    private T minHelper(Node<T> current){
        if(current==null){
            return null;
        }

        if(current.left == null){
            return current.element;
        } else{
            return minHelper(current.left);
        }
    }

    @Override
    public Node<T> findNode(T element){
        if(isEmpty()){
            throw new NullPointerException("Can't find node containing "+element+", BST is Empty!");
        }
        return findNodeHelper(this.root, element);
    }

    private Node<T> findNodeHelper(Node<T> current, T element){
        if(current==null){
            return null;
        }

        if(element.compareTo(current.element)==-1){
            if(element.compareTo(current.left.element)==0){
                return current.left;
            } else{
                return findNodeHelper(current.left,element);
            }
        } else{
            if(element.compareTo(current.right.element)==0){
                return current.right;
            } else{
                return findNodeHelper(current.right,element);
            }
        }
    }

    @Override
    public Node<T> findParent(T element){
        if(isEmpty()){
            throw new NullPointerException("Can't find Parent of "+element+", BST is Empty!");
        } else if(this.root.element==element){
            throw new NullPointerException("Can't find Parent of"+element+", this is the Root Node!");
        }
        return findParentHelper(this.root,element);
    }

    private Node<T> findParentHelper(Node<T> current, T element){
        if(current==null){
            return null;
        }

        if(element.compareTo(current.element)==-1){
            if(element.compareTo(current.left.element)==0){
                return current;
            } else{
                return findParentHelper(current.left,element);
            }
        } else{
            if(element.compareTo(current.right.element)==0){
                return current;
            } else{
                return findParentHelper(current.right,element);
            }
        }
    }

}

File: Python/Data Structures & Algorithms Module - Python/Revision/BST1/BST.py
"""
BST.py
Billy Cussen
26/01/2021
"""

from Node import Node

class BST:

    def __init__(self):
        self.root = None

    def isEmpty(self):
        return self.root == None

    def size(self):
        if self.isEmpty():
            return 0
        return self.sizeHelper(self.root)

    def sizeHelper(self, current):
        if current == None:
            return 0
        return 1 + self.sizeHelper(current.left)+self.sizeHelper(current.right)

    def depth(self):
        if self.isEmpty():
            return 0
        return self.depthHelper(self.root)

    def depthHelper(self, current):
        if current == None:
            return 0

        leftDepth = 1 + self.depthHelper(current.left)
        rightDepth = 1 + self.depthHelper(current.right)
        if leftDepth >= rightDepth:
            return leftDepth
        else:
            return rightDepth

    def min(self):
        if self.isEmpty():
            return None
        return self.minHelper(self.root)

    def minHelper(self, current):
        if current.left == None:
            return current.element
        else:
            return self.minHelper(current.left)

    def max(self):
        if (self.isEmpty()):
            return None
        return self.maxHelper(self.root)

    def maxHelper(self, current):
        if current.right == None:
            return current.element
        else:
            return self.maxHelper(current.right)

    def contains(self, element):
        if(self.isEmpty()):
            return False
        return self.containsHelper(self.root, element)

    def containsHelper(self, current, element):
        if current == None:
            return False
        
        if ((element > current.element) - (element < current.element) == 0):
            return True
        elif ((element > current.element) - (element < current.element) <= -1):
            return self.containsHelper(current.left, element)
        else:
            return self.containsHelper(current.right, element)

    def insert(self, element):
        if self.isEmpty():
            self.root = Node(element)
        else:
            self.insertHelper(element, self.root)

    def insertHelper(self, element, current):
        if((element > current.element)-(element < current.element)<=-1):
            if current.left == None:
                current.left = Node(element)
            else:
                self.insertHelper(element, current.left)
        else:
            if current.right == None:
                current.right = Node(element)
            else:
                self.insertHelper(element, current.right)


    def inOrder(self):
        if self.isEmpty():
            print("Can't do inorder traversal, BST is Empty!")
        else:
            print("***BST INORDER TRAVERSAL***")
            self.inOrderHelper(self.root)
            print("***END***")   

    def inOrderHelper(self, current):
        if current == None:
            return
        
        self.inOrderHelper(current.left)
        print(current.toString())
        self.inOrderHelper(current.right)

    def getNode(self, element):
        if self.isEmpty():
            return None
        else:
            return self.getNodeHelper(element, self.root)

    def getNodeHelper(self, element, current):
        if current == None:
            return None

        if (element > current.element) - (element < current.element) == 0:
            return current
        elif (element > current.element) - (element < current.element) == -1:
            return self.getNodeHelper(element, current.left)
        else:
            return self.getNodeHelper(element, current.right)

    def getParentNode(self, element):
        if(self.isEmpty() or self.root.element == element):
            return None
        else:
            return self.getParentNodeHelper(element, self.root)

    def getParentNodeHelper(self, element, current):
        if (element > current.element)-(element < current.element) == -1:
            if (element > current.left.element) - (element < current.left.element) == 0:
                return current
            else:
                return self.getParentNodeHelper(element, current.left)
        else:
            if (element > current.right.element) - (element < current.right.element) == 0:
                return current
            else:
                return self.getParentNodeHelper(element, current.right)